### PR TITLE
feat: added support for device mode transformations

### PR DESCRIPTION
**Pull request by desusai7**
## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
## Checklist:
- [ ] Version updated in `README`
- [ ] Version updated in `RSConstants.m`(v1)/`RSConstant.swift`(v2)
- [ ] Version updated in `Rudder.xcodeproj`
- [ ] Version updated in `Rudder.podspec`
- [ ] `CHANGELOG` Updated
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation
- [ ] Passed `pod lib lint --no-clean --allow-warnings`

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-ios/160)

### CODE REVIEW

**Review by pallabmaiti — COMMENTED** [review]
Please remove the `.DS_Store` files.

File: Sources/Classes/RSApplicationLifeCycleManager.m [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        self->config = config;
+        self->preferenceManager = preferenceManager;
+        self->firstForeGround = YES;
+        self->backGroundModeManager = backGroundModeManager;
+    }
+    return self;
+}
+
+#if !TARGET_OS_WATCH
+- (void) trackApplicationLifeCycle {
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/RSApplicationLifeCycleManager.m:28]: Yes, can do that but isn't the code more readable and maintainable this way ?
Also, this wouldn't lead any issues as the code is conditionally compiled for watchOS as well as other platforms

File: Sources/Classes/RSNetworkManager.m [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+    
+    int __block respStatus = NETWORKSUCCESS;
+    NSString *requestEndPoint = nil;
+    switch(endpoint) {
+        case TRANSFORM_ENDPOINT:
+            requestEndPoint = [self->config.dataPlaneUrl stringByAppendingString:@"/transform"];
+            break;
+        default:
+            requestEndPoint = [@"https://e582-2409-4070-2e8f-e60d-94ce-840b-d457-d541.ngrok.io" stringByAppendingString:@"/v1/batch"];
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/RSNetworkManager.m:42]: Okay, that got accidentally pushed. And it's invalidated now. Will make sure to not push them again

File: Sources/Classes/RSDeviceModeManager.m [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    }
+}
+
+- (void) initiateCustomFactories {
+    if (self->config == nil || config.customFactories == nil || config.customFactories.count == 0) {
+        [RSLogger logInfo:@"EventRepository: initiateCustomFactories: No custom factory found"];
+        return;
+    }
+    for (id<RSIntegrationFactory> factory in self->config.customFactories) {
+        @try {
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/RSDeviceModeManager.m:84]: So, sometimes few destination factories are throwing exceptions while processing a message, and in that case that particular message is not dumped to the subsequent factories in the queue, so this approach would ensure that each event is delivered to all the factories, even if one of the factories throw an exception while processing the event

File: Sources/Classes/RSDeviceModeManager.m [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            NSMutableDictionary<NSString*, RSServerDestination*> *destinationDict = [[NSMutableDictionary alloc] init];
+            for (RSServerDestination *destination in destinations) {
+                [destinationDict setObject:destination forKey:destination.destinationDefinition.displayName];
+            }
+            for (id<RSIntegrationFactory> factory in self->config.factories) {
+                RSServerDestination *destination = [destinationDict objectForKey:factory.key];
+                if (destination != nil && destination.isDestinationEnabled == YES) {
+                    NSDictionary *destinationConfig = destination.destinationConfig;
+                    if (destinationConfig != nil) {
+                        @try {
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/RSDeviceModeManager.m:62]: So, if the initialization of one factory got aborted due to any exception, then the initialization of all factories is being aborted, as we are not catching that exception here

File: Sources/Classes/RSDBPersistentManager.m [hunk 5]
```diff
@@ -27,155 +37,260 @@ - (void)createDB {
...
-    if (sqlite3_prepare_v2(self->_database, querySQL, -1, &queryStmt, nil) == SQLITE_OK) {
-        [RSLogger logDebug:@"events fetched from DB"];
-        while (sqlite3_step(queryStmt) == SQLITE_ROW) {
-            int messageId = sqlite3_column_int(queryStmt, 0);
-            const unsigned char* queryResultCol1 = sqlite3_column_text(queryStmt, 1);
-            NSString *message = [[NSString alloc] initWithUTF8String:(char *)queryResultCol1];
-            [messageIds addObject:[[NSString alloc] initWithFormat:@"%d", messageId]];
-            [messages addObject:message];
+    
+    @synchronized (DB_LOCK) {
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/RSDBPersistentManager.m:190]: Sure, will make changes accordingly

File: Sources/Classes/Public/RSNetworkManager.h [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    NETWORKSUCCESS =0,
+    WRONGWRITEKEY =2
+} NETWORKSTATE;
+
+typedef enum {
+    BATCH_ENDPOINT = 0,
+    TRANSFORM_ENDPOINT = 1
+} ENDPOINT;
+
+NSString* const STATUS = @"STATUS";
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/Public/RSNetworkManager.h:26]: Yes, I can move that, but these constants are not used anywhere else and are very specific to this implementation itself and hence kept them here. WDYT ?

File: Sources/Classes/RSApplicationLifeCycleManager.m (hunk 1 above)
[the remark this answers is not in the document]
Comment (pallabmaiti) [reply Sources/Classes/RSApplicationLifeCycleManager.m:28]: Okay. No problem.

File: Sources/Classes/Public/RSNetworkManager.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (pallabmaiti) [reply Sources/Classes/Public/RSNetworkManager.h:26]: Okay. Keep it.

File: Sources/Classes/RSDeviceModeManager.m (hunk 4 above)
[the remark this answers is not in the document]
Comment (pallabmaiti) [reply Sources/Classes/RSDeviceModeManager.m:62]: Let's sit together for this.

[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/RSDeviceModeManager.m:62]: Sure

[the remark this answers is not in the document]
Comment (pallabmaiti) [reply Sources/Classes/RSDeviceModeManager.m:70]: Explained over a call.

File: Sources/Classes/RSDeviceModeManager.m (hunk 3 above)
[the remark this answers is not in the document]
Comment (pallabmaiti) [reply Sources/Classes/RSDeviceModeManager.m:92]: Explained over a call.

**Comment by pallabmaiti** [comment]
> Please remove the `.DS_Store` files. Add this to `.gitignore`.

I can see, still 1 file added.

File: Tests/ServerConfigManagerTests.swift [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      }
+    ]
+  }
+}
+"""
+        rsPreferenceManager = RSPreferenceManager()
+        rsPreferenceManager.saveConfigJson(configJSON)
+        
+        let rsConfig = RSConfigBuilder().withControlPlaneUrl("https://invalid-rudder.com").build()
+        let rsNetworkManager = RSNetworkManager.init(config: rsConfig, andAuthToken: ServerConfigManagerTests.authToken, andAnonymousIdToken:ServerConfigManagerTests.anonymousIdToken)
```
Comment (pallabmaiti) [inline Tests/ServerConfigManagerTests.swift:122]: Missing parameter `dataResidencyManager`

File: Tests/ServerConfigManagerTests.swift [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        rsServerConfigManager = RSServerConfigManager.init(ServerConfigManagerTests.writeKey, rudderConfig: rsConfig, andNetworkManager: rsNetworkManager!)
```
Comment (pallabmaiti) [inline Tests/ServerConfigManagerTests.swift:123]:
```suggestion
        rsServerConfigManager = RSServerConfigManager(ServerConfigManagerTests.writeKey, rudderConfig: rsConfig, andNetworkManager: rsNetworkManager!)
```

File: Sources/Classes/RSContext.m [hunk 9]
```diff
@@ -67,6 +67,24 @@ - (instancetype)init {
...
 }
 
+- (instancetype) initWithDict:(NSDictionary*) dict {
+    self = [super init];
+    if(self) {
+        _app = [[RSApp alloc] initWithDict:dict[@"app"]];
+        _traits = dict[@"traits"];
+        _library = [[RSLibraryInfo alloc] initWithDict:dict[@"library"]];
+        _os = [[RSOSInfo alloc] initWithDict:dict[@"os"]];
+        _screen = [[RSScreenInfo alloc] initWithDict:dict[@"screen"]]; //
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/RSContext.m]: done

File: Sources/Classes/RSNetworkManager.m [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+            return @"v1/batch";
+        case TRANSFORM_ENDPOINT:
+            return @"transform";
+        case SOURCE_CONFIG_ENDPOINT:
+            return [[NSString alloc] initWithFormat:@"sourceConfig?p=ios&v=%@", RS_VERSION];
+    }
+}
+
+- (void) updateCTSAuthToken {
+    self->ctsAuthToken = [[RSPreferenceManager getInstance] getAuthToken];
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/RSNetworkManager.m]: removed it

File: Sources/Classes/Headers/Public/RSDBMessage.h [hunk 11]
```diff
@@ -13,6 +13,7 @@ NS_ASSUME_NONNULL_BEGIN
 @interface RSDBMessage : NSObject
     @property (nonatomic) NSMutableArray<NSString *>* messages;
     @property (nonatomic) NSMutableArray<NSString *>* messageIds;
+@property (nonatomic) NSMutableArray<NSNumber *>* statuses;
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/Headers/Public/RSDBMessage.h]: updated

File: Sources/Classes/Headers/Public/RSNetworkEnums.h
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/Headers/Public/RSNetworkEnums.h:1]: moved

File: Sources/Classes/Headers/Public/RSUtils.h [hunk 12]
```diff
@@ -13,6 +13,11 @@
 
 NS_ASSUME_NONNULL_BEGIN
 
+typedef enum {
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/Headers/Public/RSUtils.h]: moved

File: Sources/Classes/Headers/Public/RSDBPersistentManager.h [hunk 13]
```diff
@@ -11,20 +11,40 @@
 #import "RSDBMessage.h"
 #import "RSUtils.h"
 
+typedef enum {
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Sources/Classes/Headers/Public/RSDBPersistentManager.h]: moved

File: Podfile [hunk 14]
```diff
@@ -12,10 +12,23 @@ project 'Examples/RudderSammpleAppSwift/RudderSammpleAppSwift.xcodeproj'
 project 'Examples/RudderSampleApptvOSObjC/RudderSampleApptvOSObjC.xcodeproj'
 project 'Examples/RudderSampleAppwatchOSObjC/RudderSampleAppwatchOSObjC.xcodeproj'
 
+target 'Rudder' do
```
[the remark this answers is not in the document]
Comment (desusai7) [reply Podfile]: removed

File: Tests/ServerConfigManagerTests.swift (hunk 7 above)
Comment (desusai7) [reply Tests/ServerConfigManagerTests.swift:128]: added it

File: Tests/ServerConfigManagerTests.swift (hunk 8 above)
Comment (desusai7) [reply Tests/ServerConfigManagerTests.swift:129]: updated

File: Sources/Classes/RSDeviceModeManager.m [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+- (void) dumpOriginalEvents:(NSArray *) originalPayloads {
+    [RSLogger logWarn:@"RSDeviceModeManager: dumpOriginalEvents: Dumping the original events back to the transformation enabled destinations as the transformations feature is not enabled"];
+    for(NSDictionary* originalPayload in originalPayloads) {
+        RSMessage* originalMessage = [[RSMessage alloc] initWithDict:originalPayload[@"event"]];
+        NSArray<NSString *> * transformationEnabledDestinations = [self getDestinationsWithTransformationStatus:ENABLED fromMessage:originalMessage];
+        [self dumpEvent:originalMessage toDestinations:transformationEnabledDestinations withLogTag:@"dumpOriginalEvents"];
+    }
+}
+
+-(void) dumpTransformedEvents:(NSArray*) transformedPayloads ToDestination:(NSString*) destinationId {
```
[the remark this answers is not in the document]
Comment (pallabmaiti) [reply Sources/Classes/RSDeviceModeManager.m]: The first letter of `toDestinationId` is still as capital.

**Review by pallabmaiti — APPROVED** [review]

**Pull request merged**